beater/otcollector.go

```diff
@@ -0,0 +1,141 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package beater
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/open-telemetry/opentelemetry-collector/config/configmodels"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/receiver"
+	"github.com/open-telemetry/opentelemetry-collector/receiver/jaegerreceiver"
+	"github.com/pkg/errors"
+	"go.elastic.co/apm"
+
+	"github.com/elastic/apm-server/beater/config"
+	"github.com/elastic/apm-server/model"
+	"github.com/elastic/apm-server/publish"
+	"github.com/elastic/apm-server/transform"
+)
+
+type otcollector struct {
+	logger         *logp.Logger
+	traceReceivers []receiver.TraceReceiver
+	host           *host
+}
+
+func newOTCollectors(logger *logp.Logger, cfg *config.Config, _ *apm.Tracer, reporter publish.Reporter) (*otcollector, error) {
+	traceConsumer := &Consumer{
+		TransformConfig: transform.Config{},
+		ModelConfig:     model.Config{Experimental: cfg.Mode == config.ModeExperimental},
+		Reporter:        reporter,
+	}
+
+	ctx := context.Background()
+	host := newHost(ctx)
+
+	factory := &jaegerreceiver.Factory{}
+	//TODO(simi): make port configurable
+	receiverCfg := &jaegerreceiver.Config{
```

Comment:
It's a shame that we can't pass in `grpc.ServerOption`s via jaegerreceiver.Config, like you can through jaegerreceiver.Configuration/New. If we could, that would enable us to instrument the endpoint (using apmgrpc) as well as introduce rate-limiting if need be. Maybe we can send a PR upstream.

Reply:
Fully agree, it bothered me a lot that we have no control about the grpc server. I think we have two options here: try to get an upstream PR in that allows to add options or call `trace_receiver.New()` directly and not via the `factory`, in which case we would have to duplicate some logic. 
I would rather go with the approach of contributing back to not diverge too much.

Reply:
After looking more into what is missing, I now went down the road of not using the `jaegerreceiver` directly, but implementing the jaeger protobuf contract and starting a `grpc` server on our own. We still use some `open-telemetry` functions, but have full control over the grpc server. This is still only a draft, to see which direction we should go.
@axw please let me know what you think.